src/Voron/Impl/Journal/WriteAheadJournal.cs

```diff
@@ -624,6 +626,15 @@ public void ApplyLogsToDataFile(CancellationToken token, TimeSpan timeToWait)
                     if (_waj._env.Disposed)
                         return;
 
+                    if (_failedToUpdateJournalState)
+                    {
+                        if (_waj._logger.IsWarnEnabled)
+                        {
+                            _waj._logger.Warn("Ignoring the flush due to failure in updating the journal state. This is a catastrophic failure so the environment is going to be restarted.");
```

Comment:
What will cause this to generate catastrophic error here?

Reply:
Shouldn't we throw, rather than return?

Reply:
Please see the comment we have below:

https://github.com/ravendb/ravendb/blob/0a2f8fc1b739444faafd72fe7ab99921fe5b9ee9/src/Voron/Impl/Journal/WriteAheadJournal.cs#L674-L682

So effectively we will throw and mark the environment to be in catastrophic state so it will get restarted. This is handled by `BackgroundFlushWritesToDataFile()`:

https://github.com/ravendb/ravendb/blob/0a2f8fc1b739444faafd72fe7ab99921fe5b9ee9/src/Voron/StorageEnvironment.cs#L1402-L1405

The problem is that we might have another flush scheduled already. We might execute it but we've already cleared the state:

https://github.com/ravendb/ravendb/blob/0a2f8fc1b739444faafd72fe7ab99921fe5b9ee9/src/Voron/Impl/Journal/WriteAheadJournal.cs#L690

This made that the validation I added in this PR detected some scratch pages to be not freed then. Which was correct but the point is that we must not use perform any operations on the environment since the catastrophic error.

Reply:
That being the case, I think this needs a comment explaining the logic right there.
Although - just throwing again would also be fine, no?

Reply:
Throwing would overwrite the original catastrophic error so I wanted to avoid that.

In general, I have some doubts about the error handling we have now here, with this new approach. 

There are still two issues related to freeing of scratch pages so depending on our finding we might need to adjust it anyway. I will add a comment for now.